### ISSUE TITLE
Alphabetical text sorting does not work very well, especially for not-English text (but not only)

### ISSUE THREAD

**Issue by grzegorj** [notepad-plus-plus/notepad-plus-plus#13456]
### Description of the Issue
The expected behaviour of the sort function is to operate correctly not only for English. Unfortunately, it is not the case.

### Steps to Reproduce the Issue
Open a new text file and write, one word in a line, ex.
laka
łąka
mąka
mika

(note EOL = the end of the line at the end of the text)
These are Polish words, and the lines have already been sorted correctly by the alphabetical order. Nevertheless, try to sort them in Notepad++.

### Expected Behavior
The function “Sort” should not change anything in this text

### Actual Behavior
The final EOL is transferred at the beginning of the file, and the words are not correctly sorted any longer (the correct alphabetical order is totally ignored). You receive:

laka
mika
mąka
łąka
(note the empty line at the beginning; there is no EOL after the word “łąka”)
 
The bug is checked in Notepad++ v. 8.5
BTW, the correct alphabetical order in Polish is aąbcćdeęfghijklłmnńoópqrsśtuvwxyzźż. Each letter is treated separately, not as a variant with a diacritic.

**Comment by grzegorj** [comment]
Indeed, there are two issues (but within the same scope). For me, ignoring the correct alphabetical order is much more important than the issue with EOL (which I have noticed by the way). Either Notepad++ is “totally English” (and then it can sort only in accordance to ASCII codes), or it is international. But if the other alternative is correct, the lack of a possibility to sort text lines correctly is a very serious deficiency of the program. Of the programs I know, only MS Word can do it correctly, using the system locale (which is present under Windows) instead of low-level sorting based on ASCII codes. It would be REALLY nice if also Notepad++ had a similar option. And it would not be so English-biased any longer as it is now, despite existing superficial localizations (of the menu etc.). Really. The thing is worth fixing. Even if it may not be very simple.

**Comment by rdipardo** [comment]
> Of the programs I know, only MS Word can do it correctly, using the system locale (which is present under Windows) instead of low-level sorting based on ASCII codes. It would be REALLY nice if also Notepad++ had a similar option.

Apples to oranges.

Notepad++ is a *source code* editor. Source code is almost always (for [historical reasons](https://pl.wikipedia.org/wiki/ASCII#Historia)) a plain text file of ASCII keywords and punctuation symbols.

MS Word is a *business document* editor. Businesses have offices around the world, so naturally their software can intelligently edit internationalized text.

**Comment by grzegorj** [comment]
> Notepad++ is a source code editor. Source code is almost always (for [historical reasons](https://pl.wikipedia.org/wiki/ASCII#Historia)) a plain text file of ASCII keywords and punctuation symbols.

Not exactly.

Notepad++ is a free (as in “free speech” and also as in “free beer”) source code editor and Notepad replacement that supports several languages. (https://notepad-plus-plus.org/)

As you can read, Notepad++ is not only source code editor. It is also a Notepad replacement. And it, as if, supports several languages. It is not true: in its present shape, it cannot sort texts in several languages. So, there IS a problem with sorting, contrary to what rdipardo writes.

Besides, Notepad++ can serve Unicode texts. Programmers do not need Unicode to write codes! So, the statement that Notepad++ is only a source code editor is completely groundless.

Perhaps you use Notepad++ only to writing code. I do not, and plenty of other uses do not as well. I also use it to my _not business_ notes, and it is full consistant with the program’s declared purpose.

Moreover, the function under question is called “Sort”, and not “Sort the code lines”! Telling the truth, I cannot imagine the need of sorting code. But I do see a need to sort a text file that is not a code. And such a text needn’t be in English. So, the lack of correct sorting **is** a serious bug of the program, and there **is** a need to fix it.

In other words: a source code does not need to be sorted. If the Notepad++ was only a source code editor, it would not need the sort function at all. But if it has the function, it really should sort texts correctly, and not only in accordance with the English alphabet.

I am not a programmer and I cannot help with it, but I do not think it is a really hard thing to implement correct sorting of texts, so I cannot understand this negative argumentation. It is enough to forget ASCII codes of the characters, and to use different codes of them, instead. For a Polish text: assign “0” to space, “1” to “a” (and “A”, and also “á”, “Á”, “à”, “À”, “ä”, “Ä”, “å”, “Å” etc., as foreign letters “a” with diacritics should also be sorted in Polish text as if they were simple “a”’s), “2” to “ą” (and to “Ą”), “3” to “b” etc. And then sort the text in accordance with these codes, and not with ASCII codes. Make a similar table of codes for each language Notepad++ supports. That’s all. I do not think it would be really hard. The procedure of sorting itself should be exactly the same as it is now, except those mentioned letter codes was sorted, not ASCII codes as it is now.

Once again, if rdipardo uses Notepad++ as a source code editor, he does not need to use sorting at all. But I use it for different purposes, and I need sorting. Since the program is declared to be a Notepad replacement and not a source code editor only, please stop discuss about the motivation, and just fix the bug of the program.

Thank you in advance.

**Comment by grzegorj** [comment]
One more note to what rdicardo wrote.

As I have noticed, Notepad++ does not use a two-phase sorting procedure. In this style, first, capital and lowercase letters are treated the same, and then capital letters gain higher priority over the corresponding lowecase letters.

Instead, for example, the text:
Basia
asia
basia
Asia

is sorted to:

Asia
Basia
asia
basia
(with the mentioned transferring of the EOL sign at the beginning)
which is totally incorrect even for code writers!

The expected result of sorting is:
Asia
asia
Basia
basia

(with the option “capitals first”), or:
asia
Asia
basia
Basia

(with the option “lowercase first”).

So, Notepad++ cannot sort correctly even English text.

BTW. The correct way of sorting should take into account diacritic letters. E.g. German umlauted letters, contrary to Polish letters of the type “ą” or “ł”, must be treated as simple letters when sorting. But when two words differs only in umlaut, the plain letter must be taken first. Ex. the correct alphabetic order is: Ohr, Öhr, Ohrenarzt, ohrenbetäubend, Ohrenbläser.

In Polish, “c” and “ć” are different, but “c” and “č” are the same (“č” is not a letter of the Polish alphabet). So, the correct order is: cap, capek, Čapek, car, czysty, ćma. Notice that “c” and “č” have the same place in the alphabet sorting, but when “háček” is the only difference, the word with the accented letter must follow the word with the plain letter.

Sorting is not a simple procedure (but should really be simple to implement), and treating sorting of text lines as sorting of ASCII codes is a serious bug. Even for programmers...

**Comment by grzegorj** [comment]
I have just found a program which is a declared as a clear programmer’s editor, PsPad (http://www.pspad.com/en/). Despite it is a clear source code editor and not a text editor, it can sort texts fully correclty, and even better than MS Word does.

Anyway, compares of oranges to apples are just clear fantasies of an arogant person. We should collaborate and not admonish others and not pretend to be smart (while being nothing more but rude) and send others away to Wikipedia articles for them to learn such or another thing. PsPad is not a “business document editor” and can sort text files correctly. So, Notepad++ has a serious bug, and lacks correct sorting.

Dear developers, please correct this bug. Take an example from the other source code editor!

**Comment by molsonkiko** [comment]
@grzegorj , in spite of your obnoxious lecturing of programmers about what "should be easy to implement" despite having *not the faintest idea* of how little relationship there is between the *outside-view perceived difficulty* of programming something and the *actual difficulty* of implementing it, I happen to be interested in trying to work on this problem, or at least a more tractable subset of it.

If Notepad++ were implemented in C# or some other more modern programming language, the fix you describe would probably be only moderately obnoxious, but as it stands, it is guaranteed to be a lot of work.

**Comment by molsonkiko** [comment]
Any help trying to make my new sorter class culture-sensitive would be appreciated.
I'm pretty sure it doesn't work yet though.

**Comment by alankilborn** [comment]
>(note the empty line at the beginning; there is no EOL after the word “łąka”)

>I see what you're saying about the transposition of empty lines when you sort the file. 

@grzegorj 's complaint regarding something that is empty is that sorting (lex. ascending) this file:

![image](https://user-images.githubusercontent.com/77065706/230357007-1452f48a-f372-45b0-92c8-2fb25d4493a9.png)

_incorrectly_ results in this:

![image](https://user-images.githubusercontent.com/77065706/230357226-005eeb68-b079-4abb-ba17-27facfbb37a3.png)

How is that in any way correct?
But, stunningly, the author of Notepad++ judged it so. :-(

Further, and also stunning, if one does Ctrl+a before the sort, to select all text and every line in the file:

![image](https://user-images.githubusercontent.com/77065706/230358152-03b320c6-0899-49bf-a838-500dea461afa.png)

Then the sort result _IS_ correct:

![image](https://user-images.githubusercontent.com/77065706/230358409-b81c7a5c-a659-4344-934c-5ee38afacd8d.png)

>Probably the easiest solution would simply be to remove all empty lines

Also, if one does _Remove Empty Lines_, pre-sort:

![image](https://user-images.githubusercontent.com/77065706/230358629-307b92c7-35d0-422c-9ba0-a0ec8ead4208.png)

And then sorts, a reasonable result is obtained:

![image](https://user-images.githubusercontent.com/77065706/230358756-fffe51bf-0a21-4671-b78a-d698ebce387f.png)

But, as I don't consider the last "line" as a true line unless it has a line-ending on it (and I enforce this, for all of my files, with the `editorconfig` plugin), I can't obtain the sort result I want without doing a lot of "pre-thinking" about how to obtain correct results.

**_THIS_** is the part that I wish @grzegorj  had broken out into a completely separate issue when I said:

>There are TWO distinct issues here.

Combining it with a language-specific sorting order problem completely loses this detail.

**Comment by molsonkiko** [comment]
Good point.

It looks like there's some logic in `NumericSorter` in [Sorters.h](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/6326340ccb874af3374e95cd3bd0bd630b454b25/PowerEditor/src/MISC/Common/Sorters.h#L527) that could be changed slightly, moved to a separate function, and called before sorting in each of the ISorter implementations in that file.

**Comment by alankilborn** [comment]
> It looks like there's some logic for removing empty lines in `NumericSorter` ...

No empty lines need to be removed.
Just more care need to be taken into account to avoid that empty non-line at the end of file, rather than letting it get drug into the search (and thus "sorted" to the top of the results).

**Comment by molsonkiko** [comment]
maybe changing [this function](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/6326340ccb874af3374e95cd3bd0bd630b454b25/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp#L3838) to the below would work?
It did for me locally, now if there's an empty line at EOF it's not moved.
The current version of this function *specifically chose* to not remove that last empty line if sorting the entire document. So basically, went out of its way to behave in a way you didn't like.

```cpp
void ScintillaEditView::sortLines(size_t fromLine, size_t toLine, ISorter* pSort)
{
	if (fromLine >= toLine)
	{
		return;
	}

	const auto startPos = execute(SCI_POSITIONFROMLINE, fromLine);
	const auto endPos = execute(SCI_POSITIONFROMLINE, toLine) + execute(SCI_LINELENGTH, toLine);
	const generic_string text = getGenericTextAsString(startPos, endPos);
	std::vector<generic_string> splitText = stringSplit(text, getEOLString());
	bool lastLineEmpty = splitText.rbegin()->empty();
	if (lastLineEmpty)
	{
		splitText.pop_back();
	}
	assert(toLine - fromLine + 1 == splitText.size());
	const std::vector<generic_string> sortedText = pSort->sort(splitText);
	generic_string joined = stringJoin(sortedText, getEOLString());
	assert(joined.length() + getEOLString().length() == text.length());
	if (lastLineEmpty)
	{
		joined += getEOLString();
	}
	if (text != joined)
	{
		replaceTarget(joined.c_str(), startPos, endPos);
	}
}
```

**Comment by alankilborn** [comment]
@grzegorj 

A new plugin called Columns++ offers a sorting option; you may want to try it to see if it sorts your text differently/better.

See:

* https://community.notepad-plus-plus.org/topic/24353/new-plugin-columns

* https://github.com/Coises/ColumnsPlusPlus

**Comment by molsonkiko** [comment]
Second alankilborn's suggestion.
Before:
```
bass
baßk 
bLue
blue
blüe
blve
oyster
öyster
spä
spb
Spb
```
after `Columns++ -> sort descending (locale)`:
```
spb
Spb
spä
öyster
oyster
blve
blüe
bLue
blue
baßk 
bass
```

**Comment by alankilborn** [comment]
>Caveat: (at the time of this writing) you need to select the region you want to sort first, or the command does nothing.

I don't think that is a caveat ... the plugin is built around the concept of acting on a selection, specifically a column-selection.

>do people think this request is too niche to bother including as a default feature, or should it be added anyway 

I think it would be best as a core N++ feature.

**Comment by molsonkiko** [comment]
I guess the other question is: does it make more sense to simply replace the existing lexicographic sort with a locale-sensitive sort, or add a new locale-sensitive sort on top of the current lexicographic one?

**Comment by alankilborn** [comment]
>does it make more sense to simply replace the existing lexicographic sort with a locale-sensitive sort, or add a new locale-sensitive sort 

I'm sure if the new one _replaces_ the old one, someone will be impacted that they can't do the same sort they used to.  I'd suggest _adding_, not replacing.

**Comment by Coises** [comment]
Just a note, for what it might be worth:

When I added sort functions to [Columns++](https://github.com/Coises/ColumnsPlusPlus), my purpose was to deal with the fact that Notepad++ sort with a rectangular selection doesn't work when there are tabs in the file. I thought of suggesting a change in Notepad++ and offering a pull request, but I realized the entire sorting strategy used by Notepad++ would have to change. My impression is that the existing sort is meant to be reasonably efficient even for very large files. That was not a priority for me, while working with tabs was.

I came up with the "locale" sort when I started to look into exactly what I'd need to do to make a "case insensitive" sort. Outside the ASCII range, that concept isn't well-defined without specifying a locale; knowing the code page isn't good enough. That led me to the code I use [here](https://github.com/Coises/ColumnsPlusPlus/blob/master/src/Sort.cpp#L47). (At present, `options` is always `LCMAP_SORTKEY | NORM_LINGUISTIC_CASING | LINGUISTIC_IGNORECASE | SORT_DIGITSASNUMBERS`.) This method fit sensibly with what I was already doing (storing sort keys as members of objects in a vector, with the full line data stored elsewhere), but it wouldn’t mesh with the method Notepad++ uses.

I plan at some point to add a user option to select a locale other than the user default, and to change some of the other [options](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex#parameters) associated with the derivation of the sort key. Right now I *think* the sort I use is the same as Windows uses to sort filenames in Windows Explorer.

**Comment by molsonkiko** [comment]
Thanks, @Coises ! 
I was going to add lexicographic ignore-case culture-sensitive sorting as an option *in addition* to the existing case-insensitive sorting, and probably just use your code as a jumping-off point.
Unfortunately I have another PR already open for the other issue the OP raised (namely, EOF at EOL being sorted to top of file), and I can't start working on implementing this new sort until something is done on my other PR.

**Comment by alankilborn** [comment]
>I can't start working on implementing this new sort until something is done on my other PR

Why's that?

**Comment by Coises** [comment]
> > does it make more sense to simply replace the existing lexicographic sort with a locale-sensitive sort, or add a new locale-sensitive sort
> 
> I'm sure if the new one _replaces_ the old one, someone will be impacted that they can't do the same sort they used to. I'd suggest _adding_, not replacing.

My _personal_ feeling is that a case-sensitive locale sort is useless, and a case-insensitive sort that isn’t locale-aware is incoherent (though if you only use ASCII, you’d never notice). As a practical matter, there are already a lot of sorts on that menu... so while I’d lean towards replacing just the “Ignoring Case” sorts with locale-aware sorts, picking the right terminology so users don’t get confused could be challenging.

As I mentioned, though, there’s also the problem that the sorting strategy used by Notepad++ is not readily adaptable to locale-based sorting (nor to rectangular selections when there are tabs in the file).

**Comment by molsonkiko** [comment]
> > I can't start working on implementing this new sort until something is done on my other PR
> 
> Why's that?

Because I was being silly and forgot that I could just create a new branch on my fork that was independent of the branch I was using for my other PR. 😝 

I've gotten back to work on the culture sensitive locale sort, and results so far are encouraging.

@Coises , when I was studying your code in Columns++'s [Sort.cpp](https://github.com/Coises/ColumnsPlusPlus/blob/master/src/Sort.cpp), I couldn't help but notice that you use `SORT_DIGITSASNUMBERS`, which I looked up in [the relevant documentation](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex). When I tried to use it in my fork of notepad_plus_plus, though, I couldn't use it. I'm on Windows 10, and `#include <WinNls.h>` didn't make it visible. What do you think might be going on?

**Comment by Coises** [comment]
Just a guess; from WinNls.h:
```
//  Sort digits as numbers (ie: 2 comes before 10)
#if (WINVER >= _WIN32_WINNT_WIN7)
#define SORT_DIGITSASNUMBERS      0x00000008  // use digits as numbers sort method
#endif // (WINVER >= _WIN32_WINNT_WIN7)

```
Is it possible that you're compiling for a Windows version lower than that?

**Comment by molsonkiko** [comment]
> Is it possible that you're compiling for a Windows version lower than that?

In retrospect, yes, obviously. Notepad++ would definitely be compiling for Windows 7 because it's been around for so long.

**Comment by Coises** [comment]
@molsonkiko Hmmm... I see:
`<WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>`
in [Notepad++](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/visual.net/notepadPlus.vcxproj#L34) and in your fork. Is it possible you have something different on your development machine? (It is also possible that I don't know what I'm looking at...)

**Comment by rdipardo** [comment]
> I see: `<WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>` in [Notepad++](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/visual.net/notepadPlus.vcxproj#L34) and in your fork. Is it possible you have something different on your development machine? (It is also possible that I don't know what I'm looking at...)


Despite the name, it actually refers to the *SDK* version. Visual Studio's corresponding option is more explicit:

[![vs-win-sdk-ver-selection](https://i.stack.imgur.com/PRfQg.png)](https://stackoverflow.com/a/68739010)

It determines the compiler's header search path and the linker's library path.

You could think of it as the *maximum* supported Windows version, since every new feature goes into the SDK before it's released in the next build of the OS.

Only developers of bleeding-edge apps need to care to it. Notepad++ runs almost entirely on core system libraries that have barely changed in 35 years.

**Comment by molsonkiko** [comment]
@rdipardo , thanks for explaining how that works.

I'm not 100% sure what that means about the possibility of including `SORT_DIGITSASNUMBERS`, though. I assume based on this discussion that we would have to change the Notepad++ core `vcxproj` file.

**Comment by Coises** [comment]
@molsonkiko Check the setting @rdipardo illustrated _in the development environment where you couldn't compile_ using `SORT_DIGITSASNUMBERS`.

If that doesn't explain it, follow `LINGUISTIC_IGNORECASE` to WinNls.h (right-click and choose `Go To Definition`) and scroll until you see a preprocessor statement involving `WINVER`. Hover over `WINVER` and see what the tooltip says is its value. It must be at least 0x0601 for `SORT_DIGITSASNUMBERS` to work. If it's not, and the `Windows SDK Version` setting doesn't explain it... then there's something to figure out, I guess.

**Comment by molsonkiko** [comment]
Well, I tried playing with those settings, and it didn't help.

So just on a stupid whim, I just inserted `#define SORT_DIGITSASNUMBERS      0x00000008` in `Common.h`, and wouldn't you know, *it worked*.

Now `Sort Lines Lex. Asc. Culture-sensitively ignoring case` sorts
```
11
2
```
to 
```
2
11
```
Hooray! 🎉

**Comment by rdipardo** [comment]
> from WinNls.h:
>
> ```
> //  Sort digits as numbers (ie: 2 comes before 10)
> #if (WINVER >= _WIN32_WINNT_WIN7)
> #define SORT_DIGITSASNUMBERS      0x00000008  // use digits as numbers sort method
> #endif // (WINVER >= _WIN32_WINNT_WIN7)
> ```
>
> Is it possible that you're compiling for a Windows version lower than that?

As it turns out, *yes*: see how the property sheet defines `_WIN32_WINNT`:

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/c76f178534bba518f263c8120b732801e13c4916/PowerEditor/visual.net/notepadPlus.Cpp.props#L28

`_WIN32_WINNT_VISTA` is less than `_WIN32_WINNT_WIN7`, causing `#if (WINVER >= _WIN32_WINNT_WIN7)` to return `false`, so `SORT_DIGITSASNUMBERS` is never defined.

Swapping in `_WIN32_WINNT_WIN7` lets @molsonkiko's fork compile without any kludge. Adding a `#define` that might someday be duplicated is just an accident waiting to happen.
 
Since N++ [dropped Vista](https://github.com/notepad-plus-plus/notepad-plus-plus/commit/cd64850ed8c1df8e52327a80008e0e870823f822#diff-e9c84f7007d406e81d4a29b27d7fb27803975e9fe8123ad271ab24c945d3c303) in 8.4.7 (or rather, [Microsoft did](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13074#issuecomment-1445294784)), it's a good time to ask if `_WIN32_WINNT` can target Win7 instead.

**Comment by molsonkiko** [comment]
Thanks, @rdipardo ! Not sure I would have known where to look for `_WIN32_WINNT` without your help!

A reasonable compromise that doesn't require changing anything globally would be to just define `const int SORT_DIGITSASNUMBERS = 8;` *inside the function that uses it*. It should be optimized away at compile time anyway.

**Comment by donho** [comment]
@grzegorj 
> please stop discuss about the motivation, and just fix the bug of the program.

A lot of people including @rdipardo are volunteers here to help people and try to provide a better or even best code/text editor to you with no charge. So nobody owe you **"stop discuss about the motivation, and just fix the bug of the program"**.

**Comment by grzegorj8** [comment]
> A lot of people including @rdipardo are volunteers here to help people and try to provide a better or even best code/text editor to you with no charge. So nobody owe you a **"stop discuss about the motivation, and just fix the bug of the program"**.

I am very grateful to people who want to help others, and that includes NotePad++ developers. However, I refuse to be told that NOTEpad isn't for taking notes, but that it's purely a source code editor (as its name implies, bravo!). The fact that someone gives their time and skills for free does not give them the right to despise others, mock them and make fools of them.

**Comment by molsonkiko** [comment]
> I am very grateful to people who want to help others, and that includes NotePad++ developers

Pro tip: if you want people who are volunteering their time to be interested in helping you, I recommend you stop *there* and leave out the aggrieved tone.

And before you say something snide in response to my friendly advice, remember that *I am the person who is working to address your problem __in spite of your aggrieved tone.__*

**Comment by ArkadiuszMichalski** [comment]
In the case of Polish language the correct expectation from the first post is what should be returned. Until now I hadn't noticed that Notepad++ was sorting my native language incorrectly (or rather not the way I wanted it to be). We currently have a lot of sorting commands and adding more may not be acceptable.

I'm planning to make a plugin where I will put some alternative methods compared to those in Notepad++ (mainly faster variants), but this alternative sorting would certainly be useful in it.

**Comment by grzegorj8** [comment]
@ArkadiuszMichalski According to you, the sequence
laka
mika
mąka
łąka
means that Notepad++ sorts Polish words correctly? Are you kidding? Or you do not understand what “sort” means?

Oh please... stop this stupid game! Notepad++ cannot sort at all. I have reported this bug a year and a half ago and nothing has changed. The EOL bug (“one could easily be fixed”) has not been fixed as well.

Virtually nothing has happened, except for the comments of a certain insolent ignoramus who insisted that Note-pad is not for taking notes (and for this you need Word instead) and that programmers need to sort program lines, and the backtalk of a certain boor who told me to shut up because he was trying to solve “my” problem “for me”. I had enough of this primitive boor and didn't say anything more, and what has he done for a year and a half? He has done absolutely nothing!

The problem is not mine, but every user’s, and the fact that someone devotes their free time and participates in the development of the program does not give them the right to talk in this way. He can talk this way to his dog, or maybe his girlfriend, but if she’s smart, she’ll shoot him in the face and knock his teeth out for such words.

This is the first time I’ve encountered a situation where, in response to a report of a program error, the programmers respond with a personal attack on me or try to lecture me that they do something for free, so they can do whatever they want, and I have to beg them on my knees to kindly solve a bug in the program they're working on. Is this a madhouse or a forum about the Notepad++ program?

Or maybe the programmers working here simply have no idea how to write a procedure that correctly sorts text? If they don't understand that Notepad is for editing notes and other texts, and not just for writing program code, then I'm not surprised that they can't write a procedure that correctly sorts texts. I can't do that either, but at least I'm saying it straight and not telling anyone what to use or not to use a program that is, as its name suggests, for taking notes. And they’re trying to take out their frustration on me?

Gentlemen, a little more culture would be useful. And remember that if you do something for free, it's not for yourself, but for others. Instead of barking at them, listen to what they have to say and try to solve the problems they report, and don't bark at them like rabid dogs. I recommend buying a personal culture handbook and reading a bit to learn how to react to other people's comments.

**Comment by grzegorj8** [comment]
I am not used to working with people who do not respect me at all and turn to me like a dog. I also feel very uncomfortable when I am about to help people who say that the _note_book is not used to make notes, or that the order of “laka mika mąka
łąka” means (“for him”) that the words are sorted correctly – because I start to think about the intellectual levels of such people. And I do not want to do that.

Nevertheless, since absolutely no one here can write a sorting procedure, I will try to help despite the complete lack of my programming skills.

Some time ago I wrote a program in PHP which correctly sorts Polish texts. It could certainly be written differently and simpler. Nevertheless, since no one can even be useful to someone.

`<?php

//Test polskich znaków

// Deklaracje stałych
$n=chr(13).chr(10);
define("NL",$n);
define("SP",' ');
define("TB",chr(9));

// Zmienną $sledz ustawiamy na FALSE, ale na TRUE dla debugowania programu
$sledz=TRUE;

// Inicjujemy dwuwymiarową tablicę tlinia, inaczej wyskoczy błąd przy pierwszym wywołaniu funkcji in_array
$tlinia[0][0]=TB; $tlinia[1][0]=' ';

// Dla celów debugowania otwieramy plik debug.txt
if ($sledz)
  {
  $dplik='debug.txt';
  $plik2=fopen($dplik,'w') or exit("Otwarcie pliku $dplik było niemożliwe<br />");
  flock($plik2, LOCK_EX);
  }

// Otwieramy plik alfabetu al.txt
$nplik='al.txt';
$plik=fopen($nplik,'r') or exit("Otwarcie pliku $nplik było niemożliwe<br />");
flock($plik, LOCK_SH);

// Wczytujemy kolejne niepuste linie aż do końca pliku
// Znaki podane w pierwszej linii mają taką samą pozycję w alfabecie jak spacja
$i=0;
while (!feof($plik))
	{
	$linia=trim(fgets($plik, 512));

	// Interesują nas tylko linie zawierające jakąkolwiek zawartość,
	// oraz linia pierwsza i druga (nawet gdy są puste)
	if ((strlen($linia) > 0)||($i==0)||($i==1))
		{
		// Dwuwymiarowa tablica $tlinia przechowuje wszystkie znaki alfabetu
		// Pierwszy indeks oznacza numer znaku w tablicy
		// Drugi indeks określa kolejną literę mającą taką samą pozycję
		// Tylko linie niepuste
		if (strlen($linia) > 0)
		{
      $tlinia[$i]=explode(' ',$linia);
      }
		// Tabulacja i spacja nie są ujęte w pliku al.txt, dlatego...
		if ($i==0)
		{
		// dodajemy je ręcznie do tablicy, o ile ich tam nie ma
		if (!(in_array(TB,$tlinia)))
        {
        $tlinia[0][0]=TB;
        };
		if (!(in_array(' ',$tlinia)))
        {
        $tlinia[1][0]=' ';
        }
      }
		// Debugowanie: zapisujemy tablicę $tlinia do pliku debug.txt
		if ($sledz)
		{
		foreach($tlinia[$i] as $k=>$w)
		{
		fputs($plik2,'$tlinia['.$i.']['.$k.']=');
		fputs($plik2,$tlinia[$i][$k]);
	fputs($plik2,NL);
        }
      }
		$i++;
		}
	}
flock($plik, LOCK_UN);
fclose($plik);

// Koniec debugowania
if ($sledz)
  {
  flock($plik2, LOCK_UN);
  fclose($plik2);
  }

// Tworzymy tablicę $pozycja, która literom przypisuje określone pozycje alfabetu
// Funkcja count zlicza ilość elementów w tablicy
$i_maks=count($tlinia);
for ($i=0;$i<$i_maks;$i++)
	{
	$j_maks=count($tlinia[$i]);
	for ($j=0;$j<$j_maks;$j++)
		{
		$pozycja[$tlinia[$i][$j]]=$i;
		}
	}
// Ustalamy długość pola cyfrowego w reprezentacji
// posługując się logarytmem dziesiętnym ilości elementów tablicy ;-)
// zaokrąglonym w górę (funkcja ceil)
$dlugpol=ceil(log10($i_maks));
$pole='';
for ($i=0;$i<$dlugpol;$i++) $pole.='0';
$dlugpol*=(-1);

/*
Otwieramy plik danych do posortowania,
wczytujemy rekordy i tworzymy ich reprezentacje
*/

$nplik='we.txt';
$plik=fopen($nplik,'r') or exit("Otwarcie pliku $nplik było niemożliwe<br />");
flock($plik, LOCK_SH);
$i=0;
while (!feof($plik))
	{
	$linia=trim(fgets($plik, 1024));
	if (strlen($linia) > 0)
		{
		$liniads[$i]=$linia;
		$liniarp[$i]='';
		$dlugosc=strlen($linia);
		for ($j=0;$j<$dlugosc;$j++)
			{
			$znacznik=$pole.$pozycja[$linia[$j]];
			$znacznik=substr($znacznik,$dlugpol);
			$znacznik=strtr($znacznik,"0123456789","abcdefghij");
			$liniarp[$i].=$znacznik;
			}
		$i++;
		}
	}
flock($plik, LOCK_UN);
fclose($plik);

asort($liniarp);

$plik=fopen('wy.txt','w');
flock($plik, LOCK_EX);
$imaks=count($liniarp);
reset($liniarp);
for ($i=0;$i<$imaks;$i++)
	{
		fputs($plik,$liniads[key($liniarp)]);
	fputs($plik,NL);
	next($liniarp);
	}
flock($plik, LOCK_UN);
fclose($plik);
echo 'Operacja wykonana, wynik sortowania zapisano w pliku wy.txt';
?>
`

All comments are in Polish, do the use of Google Translate if interested. The program needs the Polish alphabet in a separate file:

`
﻿

a A á Á ä Ä
ą Ą
b B
c C č Č
ć Ć
d D ď Ď
e E é É ě Ě
ę Ę
f F
g G
h H
i I í Í
j J
k K
l L
ł Ł
m M
n N ň Ň
ń Ń
o O ö Ö
ó Ó
p P
q Q
r R ř Ř
s S š Š
ś Ś
t T ť Ť
u U ú Ú ü Ü
v V
w W
x X
y Y ý Ý
z Z ž Ž
ź Ź
ż Ż
0
1
2
3
4
5
6
7
8
9

`

Use it if you find it useful.

**Comment by grzegorj8** [comment]
[al.txt](https://github.com/user-attachments/files/16418062/al.txt)
[lexsort.php.txt](https://github.com/user-attachments/files/16418071/lexsort.php.txt)
The program is highly imperfect. It can only sort the whole text in a file called we.txt, placed in the same folder, and saves the result in the file wy.txt. As for now, it can serve the Polish language only. The sorting process uses two steps. In the first step, lines are sorted according to the position of a given character in the alphabet. “A” and “a”, for example, are treated the same. Only when there are more similar lines, the second step starts. During it, “a” is sorted before “A”.

The program can be applied to other languages but it can need further modifications. The problem is that e.g. the German ß should be treated exactly as the “ss” sequence on the first step (but not on the other), so one character as two characters. There also exist languages in which two characters (like “ch”, “dz” etc.) should be treated as one character when sorting.

**Comment by grzegorj8** [comment]
Oh, thank you! I have not known about this plugin. It works OK with a Polish text, splendid work! Really, I'm not that demanding... for me, it works fine.

So perhaps it is really time to transfer the function of this plugin into the program itself. But first the EOL bug (I mentioned at the beginning) should be fixed (it is still intact after a year and a half), and the column mode could be removed (I am not sure what it is for, but it is not very harmful).

I will do more tests in a free time, and if I don't let you know, it will mean that everything is fine, OK?

Thanks again!

**Comment by alankilborn** [comment]
> But first the EOL bug (I mentioned at the beginning) should be fixed (it is still intact after a year and a half),

If this is what I _think_ it is, the author has stated several times he does not feel it is a bug.  And that is the end of the story; it won't get "fixed" if the author doesn't agree with it.

**Comment by grzegorj8** [comment]
It seems that (at least in v8.6.7) the bug has been fixed and it is not present if you select lines you want to sort. But if you do not select anything, and choose the sort option, the last line loses the EOL (13h 10h) which is transferred at the beginning of the file.

Try the sequence I spoke about at the beginning:
laka
łąka
mąka
mika

(with ENTER at the end)

If you select all the text and next choose Sort, the program will sort the text by ASCII codes. Then do not select the text at all, and then try to sort it. The first line will become the second one, and the last one will lose its EOL.

Is it normal and expected to move EOL from the end to the beginning? Ooops...

**Comment by alankilborn** [comment]
> Is it normal and expected to move EOL from the end to the beginning?

This is what the author wants, although I have said a few times it is really wrong behavior.
Perhaps with another person (you @grzegorj8) really complaining about it, the author will be influenced to change his mind about it.

**Comment by alankilborn** [comment]
Another request: https://community.notepad-plus-plus.org/topic/26108

**Comment by liyanggang** [comment]
> ### Description of the Issue
> The expected behaviour of the sort function is to operate correctly not only for English. Unfortunately, it is not the case.
> 
> ### Steps to Reproduce the Issue
> Open a new text file and write, one word in a line, ex. laka łąka mąka mika
> 
> (note EOL = the end of the line at the end of the text) These are Polish words, and the lines have already been sorted correctly by the alphabetical order. Nevertheless, try to sort them in Notepad++.
> 
> ### Expected Behavior
> The function “Sort” should not change anything in this text
> 
> ### Actual Behavior
> The final EOL is transferred at the beginning of the file, and the words are not correctly sorted any longer (the correct alphabetical order is totally ignored). You receive:
> 
> laka mika mąka łąka (note the empty line at the beginning; there is no EOL after the word “łąka”)
> 
> The bug is checked in Notepad++ v. 8.5 BTW, the correct alphabetical order in Polish is aąbcćdeęfghijklłmnńoópqrsśtuvwxyzźż. Each letter is treated separately, not as a variant with a diacritic.

I found that sorting multilingual file names in "Folder as Workspace" on v8.7.9 is normal. I hope developers can apply the same sorting method to "Edit ->Line Operations ->Sort Lines Lexicographically Ascending"

![Image](https://github.com/user-attachments/assets/09b3cd37-dba1-41dc-bdd0-7eb798cfae00)

**Comment by freezer2022** [comment]
As the current issue turned out to affect many languages, and serves as the main collective / aggregate thread, wherein the most of discussion goes on, here are the other related threads:
- (10.2015) https://github.com/notepad-plus-plus/notepad-plus-plus/issues/927
- (06.2020) https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8481
- (02.2022) https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11261
- (04.2025) https://github.com/notepad-plus-plus/notepad-plus-plus/issues/16406
- (04.2025) https://github.com/notepad-plus-plus/notepad-plus-plus/issues/16409

> alankilborn (https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13456#issuecomment-1501152917): A new plugin called Columns++ offers a sorting option; you may want to try it to see if it sorts your text differently/better

<br>

By the way - speaking of plugins - I've checked and it has turned out that 2 other known Notepad++ sorting plugins also suffer same / similar problems: 
- TextFX2 (menu: _TextFX2 ==> TextFX2 Tools ==> Sort_)
- Linesort ( https://community.notepad-plus-plus.org/post/90037 and [2770 (comment)](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/2770#issuecomment-1771282307) )

**Comment by Coises** [comment]
If I might presume to summarize the situation regarding implementation of a solution to this:

There is a standard, reliable way to get a linguistically correct sort on Windows, but there are problems:

1. The sort is dependent on the locale. Some languages even have more than one sort depending on the purpose (such as German dictionary versus phonebook). Notepad++ could default to the current (system) locale, but it should be possible to override that. In my opinion (having implemented this in Columns++), doing it right would require some significant additional user interface elements (probably as a “Sorting” page in the Preferences dialog).

2. Windows locale-based sorts don’t work the way people who mostly use ASCII expect sorts to work — especially case sensitive sorts.

3. The method Notepad++ uses to implement sorting can’t be easily modified to use locale sorts efficiently. (You wouldn’t want to call [LCMapStringEx](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex) multiple times per line, so you’d need to store sort keys independently from the lines, rather than indexing into lines as the current sort does.) So there would be a lot of code change, potentially introducing new bugs in things that used to work. It is unknown (to me) what the efficiency implications would be for sorts on very large files, especially in limited-memory environments.

On the other hand:

1. This problem — apparently a source of considerable frustration and disappointment to users of languages other than English — would finally go away.

2. Problem 2 above could probably be solved in the same user interface as problem 1 by adding something like a “raw byte values” option. Implementation would almost certainly be trivial, since it would just mean bypassing the Windows call ([LCMapStringEx](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex)) that creates a locale sort key.

3. If the method I use in Columns++ were adopted, the current limitation that all line endings must be the same, and that a rectangular selection specifying the sort keys won’t work if any lines have tabs to the left of the selection, would no longer apply.

4. We could address the unexpected relocation of a line ending at the end of the file when there is no selection, which is almost never something a user would want, at the same time — since settings will be necessary anyway, we could add a checkbox if we really think there are cases where users would want sort with nothing selected to behave differently than sort after Select All.

@donho:

Considering the above, would you be open to a proposal to replace the current sort with one that recognizes locales?

I could work on the actual sorting — having implemented essentially what is required in Columns++, I believe I understand how to do it. I have not looked at the code that manages the Preferences dialog, so I’m less certain of my ability to make a good proposal for that, though having some way of specifying details for the sort would be essential.

An alternative would be to leave the existing sorts alone and add an additional “Custom Sort...” option that would open a dialog where all the details could be specified. Though this strikes me as somewhat clumsy and “ugly,” the Preferences dialog is particularly annoying to have to access often; so for people who work in multiple languages and need the appropriate sorts for each, a “Custom Sort...” dialog might actually be better. It is unclear to me which approach would be best.

**Comment by donho** [comment]
@Coises 
Good analysis.
Considering the potential problems you mentioned above, I suggest keeping the old sorting methods untouched, but adding new locale sorting functions, so the menu will look like:

```
-------------------------------------
Sort Lines Lexicographically Ascending    
Sort Lines Lex. Ascending Ignoring Case   
Sort Lines Lexicographically Ascending (Locale sort)
Sort Lines Lex. Ascending Ignoring Case (Locale sort) 
Sort Lines As Integers Ascending          
Sort Lines As Decimals (Comma) Ascending  
Sort Lines As Decimals (Dot) Ascending    
---------------------------------------
Sort Lines Lexicographically Descending   
Sort Lines Lex. Descending Ignoring Case  
Sort Lines Lexicographically Descending (Locale sort) 
Sort Lines Lex. Descending Ignoring Case (Locale sort) 
Sort Lines As Integers Descending         
Sort Lines As Decimals (Comma) Descending 
Sort Lines As Decimals (Dot) Descending   
```

If locale sorting functions are stable without issue, they can replace the old ones in the future.
What do you think?

**Comment by Coises** [comment]
@donho:
> What do you think?

I would have only one ascending and one descending Locale sort.

There is a case-sensitivity flag available when generating locale-based sort keys (more than one, in fact; I use LINGUISTIC_IGNORECASE for case-insensitive sorts), but case is the lowest-level portion of the generated key; in practice, there is almost no difference between case sensitive and case insensitive locale-based sorts. As an example, consider this text:
```
aaA
Aaa
aaa
abB
abb
aBb
ABC
```

If you sort that case insensitive, it doesn’t change. If you sort it using a case sensitive locale sort, you get:
```
aaa
aaA
Aaa
abb
abB
aBb
ABC
```

I would leave case sensitivity as an option in the locale sort settings. The only difference it makes is whether sort keys that are identical except for case will be left in the original order or sorted according to the first position in which case differs, with lower case first.

Do you think it would be best to include options as a new page in Preferences? The Locale sort details section of the Columns++ Sort dialog:

![Image](https://github.com/user-attachments/assets/d3fb2c71-fc3b-4fb0-8cfa-07c3ed46298f)

provides an example of what I think is needed. I chose [the checkbox options](https://github.com/Coises/ColumnsPlusPlus/blob/d8e2c8bcd9b753852f80422fabae6890e5418003/src/Sort.cpp#L161) from the flags available for [LCMapStringEx](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex) when specifying LCMAP_SORTKEY.

**Comment by donho** [comment]
@Coises 
I would say: if we use dialog, then we remove all the menu command, and let the dialog do the all actions; or no dialog and we use only the menu commands. 

IMO, Lexicographically Ascending / Descending case sensitive / Ignoring case (Locale sort) cover all case.
So my preference is the 2nd one.

**Comment by RISC-VVV** [comment]
> ：
> 
> > 你觉得怎么样？
> 
> 我只有一个升序和一个降序的 Locale 排序。
> 
> 生成基于区域设置的排序键时，有一个区分大小写的标志可用（实际上不止一个;我使用 LINGUISTIC_IGNORECASE 进行不区分大小写的排序），但 case 是生成的键的最低级部分;在实践中，区分大小写和不区分大小写的基于 locale 的排序之间几乎没有区别。例如，请考虑以下文本：
> 
> ```
> aaA
> Aaa
> aaa
> abB
> abb
> aBb
> ABC
> ```
> 
> 如果对不区分大小写的排序，它不会更改。如果使用区分大小写的区域设置排序对其进行排序，则会得到：
> 
> ```
> aaa
> aaA
> Aaa
> abb
> abB
> aBb
> ABC
> ```
> 
> 我会将区分大小写保留为区域设置中的一个选项。唯一的区别是，除大小写外相同的排序键是按原始顺序保留，还是根据大小写不同的第一个位置排序，小写在前。
> 
> 您认为将选项作为新页面包含在 Preferences 中是最好的吗？Columns++ Sort 对话框的 Locale sort details 部分：
> 
> ![Image](https://github.com/user-attachments/assets/d3fb2c71-fc3b-4fb0-8cfa-07c3ed46298f)
> 
> 提供了我认为需要的示例。在指定 [LCMapStringEx 时，我从可用于 LCMapStringEx](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex) 的标志中选择了[复选框选项](https://github.com/Coises/ColumnsPlusPlus/blob/d8e2c8bcd9b753852f80422fabae6890e5418003/src/Sort.cpp#L161)LCMAP_SORTKEY。
> 
> 注意：在 Columns++ 中，我有用于升序和降序 Locale 排序的菜单选项，这些选项使用默认选项（当前 Windows 区域设置、选中数字排序、未选中其他选项）。第一次尝试可能会使用默认选项，并将添加设置界面留到以后。

I found that sorting multilingual file names in "Folder as Workspace" on v8.7.9 is normal.

I hope developers can apply the same sorting method to "Edit -> Line Operations -> Sort Lines Lexicographically Ascending"

test sample：
laka
mika
mąka
łąka

![Image](https://github.com/user-attachments/assets/c6c0babd-831c-4e8c-b902-fe6b7c0c3615)

**Comment by Coises** [comment]
@RISC-VVV:
> I hope developers can apply the same sorting method to "Edit -> Line Operations -> Sort Lines Lexicographically Ascending"

I am working now on the “Locale sort” implementation described in previous messages in this issue. The method I’m using (already employed in [Locale sort in Columns++](https://coises.github.io/ColumnsPlusPlus/help.htm#sorting)) uses LCMapStringEx. With minor variations (depending on the flags passed), this is the same method used in sorting the tree view that displays Folder as Workspace (which uses lstrcmpi). [See Microsoft documentation.](https://learn.microsoft.com/en-us/windows/win32/intl/handling-sorting-in-your-applications)

It is unlikely that this will ever *replace* the existing lexicographic sorts, as the behavior of Locale-based sorts (particularly case sensitive sorts) is inconsistent with what users familiar with ordinary ASCII sorts expect.

**Comment by Coises** [comment]
I believe the artifacts [here](https://github.com/Coises/notepad-plus-plus/actions/runs/14677633152) are suitable for preliminary testing.

There is a problem I have not been able to resolve. In the 32-bit version, attempting to sort a large file (around 175 MB, 30k lines) has failed, but not always in the same way. Applying a Lexicographic sort to the same file results in a “bad allocation” message followed by program termination. With the new locale sort, I have at various times gotten that message, no action at all, a corrupted display (showing only a couple lines) or an unresponsive window.

Putting a try/catch around the sort routine didn’t help. The debugger isn’t much help, either, as it appears to hang in system code; the best I could narrow down is that it seems to happen after the sort routine is finished and while Scintilla is updating the display. The sort routine itself completes every time I have run it with breakpoints.

If anyone has a clue how to find the cause of this and catch it so it will produce an error message, rather than hanging or crashing Notepad++ or corrupting the in-memory copy of the file, I would appreciate the help.

**Comment by xomx** [comment]
@Coises 

> “bad allocation” message followed by program termination

This is from the N++ main unhandled exceptions "catcher", for your raised `std::bad_alloc` it is produced here:

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/48d953693ad9bd2829f1dfd36c852504f8dc4a6f/PowerEditor/src/winmain.cpp#L755-L759


> Putting a try/catch around the sort routine didn’t help.

It works ok for me. Your `std::bad_alloc` originates namely from your `SortLine` struct `sl` variable resizing:

https://github.com/Coises/notepad-plus-plus/blob/ed71d5bfba6d7451563a092f40c6f1a60a93d052/PowerEditor/src/MISC/Common/SortLocale.cpp#L154

- to try, just replace that line with:
```C++
                try
                {
                    sl.key.resize(m, 0);
                }
                catch (const std::bad_alloc&)
                {
                    ::MessageBoxA(NULL, "std::bad_alloc in sl.key.resize raised!", "SortLocale::sort", MB_OK | MB_ICONERROR | MB_APPLMODAL);
                    return false;
                }
```

**Comment by Coises** [comment]
@xomx 

Thank you for confirming the bad allocation and that it can be caught. With that and some more testing, I have a better grasp on the situation.

The file I was using for tests is (on my system, at least) right at the borderline of being too large to process. Using a much larger file consistently throws a bad allocation exception, which I can intercept. When the borderline file throws a bad allocation (what other factors determine whether or not that happens are not yet known), it is caught, too. Appropriate changes to my routine make it possible to intercept that, show a message, and proceed without damage.

The problem now is what happens when sorting that file *doesn’t* throw a bad allocation. In those cases, a breakpoint following the sort is hit without a problem; something happens later, apparently in Scintilla processing, that causes a hang. Forcing a break in the debugger during the hang yields a call stack like this (details vary, but at least the part from “PaintDC” back to the “External Code” indicator below it seems consistent) :

```
 	[External Code]	
>	notepad++.exe!Scintilla::Internal::CellBuffer::EditionEndRun(int pos) Line 1307	C++
 	notepad++.exe!Scintilla::Internal::Document::EditionEndRun(int pos) Line 448	C++
 	notepad++.exe!Scintilla::Internal::Document::GetMark(int line, bool includeChangeHistory) Line 430	C++
 	notepad++.exe!Scintilla::Internal::EditModel::GetMark(int line) Line 172	C++
 	notepad++.exe!Scintilla::Internal::MarginView::PaintOneMargin(Scintilla::Internal::Surface * surface, Scintilla::Internal::PRectangle rc, Scintilla::Internal::PRectangle rcOneMargin, const Scintilla::Internal::MarginStyle & marginStyle, const Scintilla::Internal::EditModel & model, const Scintilla::Internal::ViewStyle & vs) Line 318	C++
 	notepad++.exe!Scintilla::Internal::MarginView::PaintMargin(Scintilla::Internal::Surface * surface, int topLine, Scintilla::Internal::PRectangle rc, Scintilla::Internal::PRectangle rcMargin, const Scintilla::Internal::EditModel & model, const Scintilla::Internal::ViewStyle & vs) Line 515	C++
 	notepad++.exe!Scintilla::Internal::Editor::PaintSelMargin(Scintilla::Internal::Surface * surfaceWindow, const Scintilla::Internal::PRectangle & rc) Line 1842	C++
 	notepad++.exe!Scintilla::Internal::Editor::Paint(Scintilla::Internal::Surface * surfaceWindow, Scintilla::Internal::PRectangle rcArea) Line 1913	C++
 	notepad++.exe!Scintilla::Internal::ScintillaWin::PaintDC(HDC__ * hdc) Line 1286	C++
 	notepad++.exe!Scintilla::Internal::ScintillaWin::WndPaint() Line 1325	C++
 	notepad++.exe!Scintilla::Internal::ScintillaWin::WndProc(Scintilla::Message iMessage, unsigned int wParam, int lParam) Line 2351	C++
 	notepad++.exe!Scintilla::Internal::ScintillaWin::SWndProc(HWND__ * hWnd, unsigned int iMessage, unsigned int wParam, long lParam) Line 4150	C++
 	[External Code]	
 	[Frames below may be incorrect and/or missing, no symbols loaded for user32.dll]	
 	notepad++.exe!ScintillaEditView::scintillaNew_Proc(HWND__ * hwnd, unsigned int Message, unsigned int wParam, long lParam) Line 702	C++
 	notepad++.exe!ScintillaEditView::scintillaStatic_Proc(HWND__ * hwnd, unsigned int Message, unsigned int wParam, long lParam) Line 391	C++
 	[External Code]	
 	notepad++.exe!`anonymous namespace'::SurfaceD2D::Stadium(Scintilla::Internal::PRectangle rc, Scintilla::Internal::FillStroke fillStroke, Scintilla::Internal::Surface::Ends ends) Line 849	C++
```

I presume the problem is triggered by the very large replacement; the SCI_REPLACETARGET doesn’t encounter any problem, but updating the display appears to fail in an unpleasant way. 

Under debug, just loading the file, selecting all, copying and then pasting results in a program assertion that leads back to the destructor for `GlobalMemory in scintilla\win32\\ScintillaWin.cxx`. The existing Lexicographic sorts consistently fail with a bad allocation on another test file I have which is smaller than my “borderline” file and on which the new Locale sorts consistently succeed.

All this is 32-bit version only; I haven’t seen it in x64 (but I haven’t tried multi-gigabyte size files yet, either).

**Comment by xomx** [comment]
@Coises 

Hmm, it's evidently more complicated, now I tried another (bigger, 400 MB) txt-file in x86 and got a really weird behavior - N++ process running just disappears without any farewell (it even bypasses my additional SEH __try/__except handler). Until now I thought that any C++ std container will throw std::bad_alloc when it cannot allocate its memory ... but evidently not. Looks like that when the opened file line-count exceeds some critical threshold, N++ silently terminates on this std::vector allocation line:
https://github.com/Coises/notepad-plus-plus/blob/ed71d5bfba6d7451563a092f40c6f1a60a93d052/PowerEditor/src/MISC/Common/SortLocale.cpp#L106

Now I don't have more time for investigation, sorry, maybe next week.

**Comment by Coises** [comment]
Release artifacts [here](https://github.com/Coises/notepad-plus-plus/actions/runs/14684305351) include interception of exceptions during most of the sort process; that should catch failed allocations when performing the sort, but I intentionally left the Scintilla call which modifies the document outside the try block, so Notepad++ will catch errors there and apply its damage control process. This version shows a message box for errors and also when the sort is not performed because of an invalid selection (multiple selection or selection all on one line).

The peculiar case for 32-bit where sorting a file that’s big enough, but not too big, can yield unexpected results — either no change in the displayed document, a corrupted document or an unresponsive window — still remains. I have not identified what causes this, nor what determines which result will occur. I’ve now discovered that in some cases (maybe all) when there is no apparent change to the document, the document cannot subsequently be modified: typing, deleting, moving and other operations show no effect on the display, though it is still possible to scroll and to select text.

**Comment by molsonkiko** [comment]
@Coises 
> The peculiar case for 32-bit where sorting a file that’s big enough, but not too big, can yield unexpected results — either no change in the displayed document, a corrupted document or an unresponsive window — still remains.

I see this kind of problem problem all the time with things that have nothing to do with sorting whenever I'm using a document longer than maybe 80MB in Notepad++ 32bit.

I know it must be infuriating for a C++ programmer as experienced as you to see weird memory issues and not know what's causing them, but I think you should stop driving yourself crazy and leave the issue for another day.

**Comment by Coises** [comment]
@xomx 

Thank you again!

> an idea what's going on and what to do for detection

And you are correct. Having used the C++ interface (ScintillaCall) exclusively in my plugin, I forgot that the C interface (used in Notepad++) doesn’t raise exceptions, it just sets an error code. Indeed, in the troublesome cases, SCI_REPLACETARGET is returning with SC_STATUS_BADALLOC.

However, attempting to display a MessageBox still fails, at least when running within Visual Studio. If I’m interpreting what I’m seeing correctly, at some point Scintilla still tries to update its display (which would appear behind the MessageBox) and hangs while doing that, so the MessageBox is never shown. I’m still looking for a way to get some sort of sane error recovery.

I will see if raising an exception and leaving it to Notepad++ to handle works any better.

@molsonkiko 

I agree that the reasonable answer to bugs when opening large files in 32-bit Notepad++ is, “Well, don’t do that!” I accept that it can fail; I just don’t like that it fails with a hang that makes it impossible for Notepad++ to perform its error mitigation to try to preserve unsaved data.

**Comment by Coises** [comment]
[Release artifacts here](https://github.com/Coises/notepad-plus-plus/actions/runs/14689208081) incorporate the best strategy I could find for dealing with out-of-memory conditions. (I’ve only seen them in the 32-bit version.)

(The solution was to send a [WM_SETREDRAW](https://learn.microsoft.com/en-us/windows/win32/gdi/wm-setredraw) FALSE message to Scintilla if it fails in SCN_REPACETARGET before raising an exception for Notepad++ to catch, upon which it will exit as best it can. Without the WM_SETREDRAW, Scintilla hangs trying repaint itself before the error message box can be displayed. There appears to be no way to recover more gracefully; if there is a way to restore the integrity of the Scintilla document after the attempt to replace its content with the sorted text fails, I haven’t found it. If an out-of-memory condition occurs during the sort itself, before the Scintilla update, a message box informs the user that the sort failed and the document remains intact.)

**Comment by Coises** [comment]
@RISC-VVV:

In #16406 you describe a problem with sorting Chinese characters:
```
从
吧
啊
```
which don’t sort as expected. (I believe you mention this above, too, but I can’t read the Chinese text.)

When I test this using the new Locale sort, I get the same results that you indicate are incorrect. However, my Language is English with Locale “en - United States”. When I sort in Columns++ using the menu sorts, of course the results are the same; but if I use the **Sort...** menu item and set an ascending locale sort with Language: Chinese and Locale: “zh - China (Pronunciation)” I get:
```
啊
吧
从
```
which matches your depiction of the correct sort. I notice that this is not true of some of the other variants, like “Chinese (Traditional) / zh-Hant Hong Kong SAR (Stroke Count)” — demonstrating the point that *locale-based sorts are locale-dependent*.

The consequence of that point is that I am not convinced that providing a locale sort without a way to change the locale as appropriate for different texts will suit the needs of multi-lingual users. However, I speak only English, so I really have no practical experience at all as to how important this might or might not be.

If your Windows UI language is Chinese — or if you can set it to that without causing chaos for yourself — can you verify that the test version sorts correctly in that language?

---

In case you or others don’t know: the way to get the test version is to go to [the associated Actions page](https://github.com/Coises/notepad-plus-plus/actions/runs/14689208081), click on “build_windows (Release, x64)” or “build_windows (Release, Win32)” as appropriate at the left, click the chevron to the left of “Archive Artifacts for x64 / Release” (or Wind32) and click on the “Artifact download URL” at the bottom of the section. That will download a zip file containing a `notepad++.exe` file — you have to copy that file to replace the like-named file in an installation or portable copy of Notepad++. I know it’s a bit of pain, and certainly understand if anyone does not particularly want to do all that.

**Comment by RISC-VVV** [comment]
> [@RISC-VVV](https://github.com/RISC-VVV):
> 
> > I hope developers can apply the same sorting method to "Edit -> Line Operations -> Sort Lines Lexicographically Ascending"
> 
> I am working now on the “Locale sort” implementation described in previous messages in this issue. The method I’m using (already employed in [Locale sort in Columns++](https://coises.github.io/ColumnsPlusPlus/help.htm#sorting)) uses LCMapStringEx. With minor variations (depending on the flags passed), this is the same method used in sorting the tree view that displays Folder as Workspace (which uses lstrcmpi). [See Microsoft documentation.](https://learn.microsoft.com/en-us/windows/win32/intl/handling-sorting-in-your-applications)
> 
> It is unlikely that this will ever _replace_ the existing lexicographic sorts, as the behavior of Locale-based sorts (particularly case sensitive sorts) is inconsistent with what users familiar with ordinary ASCII sorts expect.

I have used the 'Local sort' function of 'Columns+' and the sorting result is correct. If possible, all texts would be sorted according to the current configuration by default, without the need to pop up a box to confirm the configuration and 'rectangular selection' every time they are sorted. That would be perfect

**Comment by Coises** [comment]
(sigh) Yet another caveat regarding the test version. I see now that my assumption/guess as to how `_nativeLangSpeaker.messageBox` works was totally wrong, and hence the way I send back error messages from the new locale sort routine to the [calling routing](https://github.com/Coises/notepad-plus-plus/blob/cc953f83a0070b402e70be451e010b2a20c9d634/PowerEditor/src/NppCommands.cpp#L861) will not work when the language is not English.

I will fix the method (though I won’t be able to include translations) before making this into a pull request; for now, be aware that while an error box will be raised when the sort cannot be performed due to an unsupported selection (single line or multiple selection) or an out-of-memory condition, the message displayed will be a different and incorrect one if the language is other than English.

**Comment by RISC-VVV** [comment]
> （叹气）关于测试版的另一个警告。我现在看到我对工作原理的假设/猜测是完全错误的，因此当语言不是英语时，我将错误消息从新的区域设置排序例程发送回[调用例程](https://github.com/Coises/notepad-plus-plus/blob/cc953f83a0070b402e70be451e010b2a20c9d634/PowerEditor/src/NppCommands.cpp#L861)的方式将不起作用。`_nativeLangSpeaker.messageBox`
> 
> 在将其提交到拉取请求之前，我将修复该方法（尽管我无法包含翻译）;目前，请注意，虽然由于不支持的选择（单行或多选）或内存不足情况而无法执行排序时会引发错误框，但如果语言不是英语，则显示的消息将不同且不正确。
> 
> 编辑以添加：[已修复（但未翻译）](https://github.com/Coises/notepad-plus-plus/actions/runs/14696041982)

I think most of the time, sorting is done on small batches of text, and there is no need to deliberately focus on selection sorting, otherwise repeated operations will appear cumbersome. By default, full-text sorting is sufficient. If you need to sort by region, you can select and copy the text separately to the newly created text for sorting. These are all flexible and adaptable.

**Comment by alankilborn** [comment]
> I think most of the time, sorting is done on small batches of text, and there is no need to deliberately focus on selection sorting, otherwise repeated operations will appear cumbersome. By default, full-text sorting is sufficient. If you need to sort by region, you can select and copy the text separately to the newly created text for sorting.

No, it's important to be able to select some number of consecutive lines in a tab, and then sort them.

**Comment by Coises** [comment]
I wrote:

> I am not convinced that providing a locale sort without a way to change the locale as appropriate for different texts will suit the needs of multi-lingual users.

After further testing, I tentatively revise that thought.

There was an error in my code that caused the sort routines to use (I think) the current *system* locale instead of the current *user* locale. That is fixed in the [latest test versions](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374).

It appears that most locales use a standard “Unicode” sort order except for details specific to that locale — so that for most texts any locale sort will give the same result as any other. An example where sorting differs is the one @RISC-VVV posted:
```
从
吧
啊
```
which sorts as above in “English - United States” and in “Chinese (Traditional) Hong Kong SAR (Stroke Count)” but opposite in “Chinese - China (Pronunciation).”

Another example is the following:
```
ace
acha
acna
acnc
acpa
acñb
```
which sorts this way in “English - United States”:
```
ace
acha
acna
acñb
acnc
acpa
```
this way in “Spanish - Spain (International)”:
```
ace
acha
acna
acnc
acñb
acpa
```
and this way in “Spanish - Spain (Traditional)”:
```
ace
acna
acnc
acñb
acpa
acha
```

However, it turns out that it is not all *that* difficult to change the default user locale. On Windows 10, it’s found at Control Panel: Region (going through Settings just takes longer to get to the same place). The drop-down at the top can be used to change the locale. After selection, for some locales, a link to “Change sorting method” appears below the drop-down (e.g., for Spanish International/Traditional). Testing shows that these settings can be changed even while Notepad++ is open, and the next Locale sort will still use the newly-selected locale and sort order.

So it seems to me that at least as a first implementation, it will be reasonable to use the user’s default locale sort without an option to override. Most users will be fine with their default locale, and any multi-lingual users who really need specific sorting in different languages can accomplish that without *too* much anguish.

**Comment by RISC-VVV** [comment]
> [@alankilborn](https://github.com/alankilborn):
> 
> > No, it's important to be able to select some number of consecutive lines in a tab, and then sort them.
> 
> Just in case it’s not clear, [my latest test version](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374) supports that (as do all prior versions).
> 
> I should probably document exactly what this code, if incorporated in Notepad++, will do:
> 
> Two menu items, **Sort Lines In Locale Order Ascending** and **Sort Lines In Locale Order Descending**, are added to the **Edit** | **Line Operations** submenu, following the **Sort Lines Lex.** ... **Ignoring Case** items.
> 
> When nothing is selected (specifically, when there is a single stream selection and the anchor and caret are the same), the entire document is sorted. After sorting, the caret “follows” the line in which it was located to the new position of the line, and remains at the same position within the line.
> 
> When there is a single stream selection encompassing two or more lines, the lines encompassed by the selection are sorted. If the end of the selection falls at the beginning of a line (that is, immediately after the line ending of the previous line), that line is not part of the selection. After sorting, the selection is set to encompass the sorted lines.
> 
> When there is a non-empty rectangular selection encompassing two or more lines, the lines encompassed by the selection are sorted using the selected text in each line as the sort key. After sorting, the rectangular selection is restored, to the extent possible, to encompass the same selections.
> 
> When there is an empty rectangular selection or a thin selection encompassing two or more lines, the lines encompassed by the selection are sorted using the text from the location of the selection to the end of each line as the sort key. After sorting, the rectangular selection is restored, to the extent possible, to indicate the same positions in the sorted lines.
> 
> When there is a selection which does not meet one of the above conditions, a message box is shown and the sort is not performed.
> 
> All sorts are performed using the sort keys generated by [LCMapStringEx](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex). The locale is set to the default user locale and the flags are set to `LCMAP_SORTKEY | NORM_LINGUISTIC_CASING | LINGUISTIC_IGNORECASE | SORT_DIGITSASNUMBERS`. The code is written so that the locale and flags (except LCMAP_SORTKEY) can be changed easily, but there is at present no user interface for doing that.
> 
> The restriction for existing Notepad++ sorts that all line endings must be the same as the document’s default line ending does not apply to the new sorts. Mixed line endings do not interfere with sorting, and each line ending stays with the line it ends. If the document ends with a line ending, the “virtual” empty line which Scintilla shows following it is not treated as a sortable line. If the document does not end with a line ending and the last line is included in the sort, the document’s default line ending will be used to end that line (unless it sorts as the last line) and line ending of the last line after sorting will be removed.
> 
> The restriction for existing Notepad++ sorts that rectangular selections will not select the actual keys used for sorting if the selection is preceded by, or contains, different numbers of bytes on different lines (this can happen when there are tabs, Unicode characters that have different UTF-8 encoding lengths, or fonts that are not fixed space) does not apply to the new sorts.

I urgently hope to be able to trigger  **Sort Lines In Locale Order Ascending** and **Sort Lines In Locale Order Descending** through shortcut keys. Don't pop up a pop-up saying "This command requires a rectangular selection." The pop-up is too cumbersome and affects efficiency and experience. The default execution is "Enclose the entire document in a rectangular selection." This is the effect I hope for, it's really exciting!

**Comment by Coises** [comment]
@RISC-VVV:

> I urgently hope to be able to trigger Sort Lines In Locale Order Ascending and Sort Lines In Locale Order Descending through shortcut keys. Don't pop up a pop-up saying "This command requires a rectangular selection." The pop-up is too cumbersome and affects efficiency and experience. The default execution is "Enclose the entire document in a rectangular selection." This is the effect I hope for, it's really exciting!

Indeed, that is how it works in the [latest test version](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374). You can set shortcut keys. There are no prompts for rectangular selections. (That is, and will remain, a Columns++ behavior. It serves a purpose there, but it has no place in regular Notepad++ menu commands.) Menu items aren’t translated yet, so you’ll see them in English no matter what interface language you use in Notepad++.

If you want to try it now, at this moment the latest test versions are:
[here for 64-bit](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374/artifacts/3024060787)
[here for 32-bit](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374/artifacts/3024060561)
Those are zip files containing a `notepad++.exe` file; you have to extract that file and use it to replace the `notepad++.exe` file in your installation (or portable copy) of Notepad++.

**Comment by RISC-VVV** [comment]
> [@RISC-VVV](https://github.com/RISC-VVV):
> 
> > I urgently hope to be able to trigger Sort Lines In Locale Order Ascending and Sort Lines In Locale Order Descending through shortcut keys. Don't pop up a pop-up saying "This command requires a rectangular selection." The pop-up is too cumbersome and affects efficiency and experience. The default execution is "Enclose the entire document in a rectangular selection." This is the effect I hope for, it's really exciting!
> 
> Indeed, that is how it works in the [latest test version](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374). You can set shortcut keys. There are no prompts for rectangular selections. (That is, and will remain, a Columns++ behavior. It serves a purpose there, but it has no place in regular Notepad++ menu commands.) Menu items aren’t translated yet, so you’ll see them in English no matter what interface language you use in Notepad++.
> 
> If you want to try it now, at this moment the latest test versions are: [here for 64-bit](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374/artifacts/3024060787) [here for 32-bit](https://github.com/Coises/notepad-plus-plus/actions/runs/14714355374/artifacts/3024060561) Those are zip files containing a `notepad++.exe` file; you have to extract that file and use it to replace the `notepad++.exe` file in your installation (or portable copy) of Notepad++.

I have already tried it out, and this is the effect I want, perfect. I look forward to the feature appearing in the newly released version. Thank you for making this sorting function present perfectly!
